model/modeldecoder/v2/decoder.go

```diff
@@ -0,0 +1,225 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package v2
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/elastic/apm-server/decoder"
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+func init() {
+	metadataRootPool.New = func() interface{} {
+		return &metadataRoot{}
+	}
+	// metadataNoKeyPool.New = func() interface{} {
+	// 	return &metadataNoKey{}
+	// }
+}
+
+var metadataRootPool sync.Pool
+
+func fetchMetadataRoot() *metadataRoot {
+	return metadataRootPool.Get().(*metadataRoot)
+}
+func releaseMetadataRoot(m *metadataRoot) {
+	m.Reset()
+	metadataRootPool.Put(m)
+}
+
+// DecodeMetadata uses the given decoder to create the input models,
+// then runs the defined validations on the input models
+// and finally maps the values fom the input model to the given *model.Metadata instance
+//
+// DecodeMetadata should be used when the underlying byte stream does not contain the
+// `metadata` key, but only the metadata.
+func DecodeMetadata(d decoder.Decoder, out *model.Metadata) error {
+	return decode(func(m *metadataRoot) error {
+		return d.Decode(&m.Metadata)
+	}, out)
```

Comment:
Closures cause allocations, so I suggest changing `decode` to accept `funct(d decoder.Decoder, m *metadataRoot) error)`, and passing in the decoder too. The functions you pass in can then be defined at the file level.

Reply:
right, good catch!